### ISSUE TITLE
How to use godot_variant_evaluate?

### ISSUE THREAD

**Issue by DND-dou** [godotengine/godot-headers#83]
Hello,I'm doing some cool things with godotC library recently.
I use **godot_variant_evaluate** to complete the 'operator'
In the API, 'godot_variant_evaluate' is like this
`void (*godot_variant_evaluate)(godot_variant_operator p_op, const godot_variant *p_a, const godot_variant *p_b, godot_variant *r_ret, godot_bool *r_valid);`
**This is my code:**
`godot_variant A;`
`godot_variant B;`
`godot_variant ret;`
`api->godot_variant_new_real(&A,43.658);`
`api->godot_variant_new_real(&B,4.36);`
`godot_bool i=false;`
`api2->godot_variant_evaluate(GODOT_VARIANT_OP_ADD,&A,&B, &ret, &i);`

I use the **mingw64** compiler on Windows x64 bit system, which compiles normally
But when I use the DLL, Godot crashes!!!
_So does anyone know what's wrong with it, please_

**Comment by DND-dou** [comment]
`Oh, I see!`
`godot_variant_equal` is in `godot_gdnative_core_1_1_api_struct`
Now I've found the way to get the `core_1_1_api`
In the official example:
I understand how to get `godot_gdnative_core_api_struct`(as 'api')
the `core_api` means `core_1_0_api`.It has a `next` property,so,you can get `core_1_1_api` by `api->next`!!!
The correct code should look like this:
`api2=(godot_gdnative_core_1_1_api_struct*)api->next`
`api2->godot_variant_equal(...)`
Also,if you want to get `core_1_2_api`,you can use `(godot_gdnative_core_1_2_api_struct*)api2->next`
Last,you can use `sth_X_X-1_api->next` to get `sth_X_X_api`
(Don't forget to cast types)

So far, I'm not sure how it works, but this progressive design is very interesting
(´∀｀)

**Comment by DND-dou** [comment]
`Oh, I see!`
`godot_variant_equal` is in `godot_gdnative_core_1_1_api_struct`
Now I've found the way to get the `core_1_1_api`
In the official example:
I understand how to get `godot_gdnative_core_api_struct`(as 'api')
the `core_api` means `core_1_0_api`.It has a `next` property,so,you can get `core_1_1_api` by `api->next`!!!
The correct code should look like this:
`api2=(godot_gdnative_core_1_1_api_struct*)api->next`
`api2->godot_variant_equal(...)`
Also,if you want to get `core_1_2_api`,you can use `(godot_gdnative_core_1_2_api_struct*)api2->next`
Last,you can use `sth_X_X-1_api->next` to get `sth_X_X_api`
(Don't forget to cast types)

So far, I'm not sure how it works, but this progressive design is very interesting
(´∀｀)

**Comment by BastiaanOlij** [comment]
To ensure that the API is backwards compatible we can't just add new methods to existing structs. It would break any plugin that assumes the struct has different methods.

So the structs form a chain, in this case we start with `godot_gdnative_core_api_struct`, it's next pointer points to a `godot_gdnative_core_1_1_api_struct` and it's next points to a `godot_gdnative_core_1_2_api_struct`.

If you use your plugin on an older version of Godot it may only have `godot_gdnative_core_api_struct` or only `godot_gdnative_core_api_struct` and `godot_gdnative_core_1_1_api_struct` with the last supported api having a next set to `nullptr`.

The same applies to all the others.

**Comment by akien-mga** [comment]
This is documented in https://docs.godotengine.org/en/stable/tutorials/plugins/gdnative/gdnative-c-example.html so I think we can close this issue. There might be a need to improve the documentation if it's not explicit enough, though I think the new GDExtension might make most of it obsolete so it might not be worth the effort?

**Comment by DND-dou** [comment]
> This is documented in https://docs.godotengine.org/en/stable/tutorials/plugins/gdnative/gdnative-c-example.html so I think we can close this issue. There might be a need to improve the documentation if it's not explicit enough, though I think the new GDExtension might make most of it obsolete so it might not be worth the effort?



> This is documented in https://docs.godotengine.org/en/stable/tutorials/plugins/gdnative/gdnative-c-example.html so I think we can close this issue. There might be a need to improve the documentation if it's not explicit enough, though I think the new GDExtension might make most of it obsolete so it might not be worth the effort?

Thank you.(๑╹ヮ╹๑)ﾉ
 So far, I have studied gdnitive for more than half a year. 
When I first studied, I didn't understand the content of the document. 
Maybe the original English content is easy to understand, but I am Chinese, and the Chinese translation of the document about gdnitive is puzzling (I bet the people who translate the Chinese document didn't understand the working way of gdnitive)!!! 
ヽ(‘⌒´メ)ノ

I can't understand anything... 
It makes it very difficult for me to study, but I've been slowly groping. Now I can understand its working principle. 
It's really difficult for novices. These are all problems of translation. 

By the way, I've learned about gdnitive in 4.0. Almost everything is integrated into variant. I think it's good. I'm waiting for Godot headers in 4.0 ╭︎( ˙º˙)╯︎

**Comment by DND-dou** [comment]
> This is documented in https://docs.godotengine.org/en/stable/tutorials/plugins/gdnative/gdnative-c-example.html so I think we can close this issue. There might be a need to improve the documentation if it's not explicit enough, though I think the new GDExtension might make most of it obsolete so it might not be worth the effort?

Thank you.(๑╹ヮ╹๑)ﾉ
So far, I have studied gdnitive for more than half a year.
When I first studied, I didn't understand the content of the document.
Maybe the original English content is easy to understand, but I am Chinese, and the Chinese translation of the document about gdnitive is puzzling (I bet the people who translate the Chinese document didn't understand the working way of gdnitive)!!!
ヽ(‘⌒´メ)ノ

I can't understand anything...
It makes it very difficult for me to study, but I've been slowly groping. Now I can understand its working principle.
It's really difficult for novices. These are all problems of translation.

By the way, I've learned about gdnitive in 4.0.
It is much simpler. Gdnitive 1. X is a thing of the past. Obviously, it is more important to continue to improve GDExtension.
